panic: runtime error: slice bounds out of range

This seems to be the same as #171, but based on the most recent release.
<!--

Thanks for filing an issue! If this is a question or feature request, just delete
everything here and write out the request, providing as much context as you can.

-->

### What version of `dep` are you using (`dep version`)?
v0.3.1

### What `dep` command did you run?
`dep init`

```
panic: runtime error: slice bounds out of range

goroutine 20 [running]:
github.com/golang/dep/internal/gps.(*gitSource).listVersions(0xc420228000, 0xb18980, 0xc42022c000, 0xc42019b7a0, 0x410bd8, 0x10, 0x873ea0, 0xc42019b701)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/vcs_source.go:224 +0x1082
github.com/golang/dep/internal/gps.maybeGitSource.try.func1(0xb18980, 0xc42022c000, 0xb18680, 0xc4200653c0)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/maybe_source.go:94 +0x71
github.com/golang/dep/internal/gps.(*supervisor).do(0xc420067080, 0xb186c0, 0xc420012278, 0x8de8f5, 0xc, 0x1, 0xc42019b8f0, 0xc420226000, 0x0)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/source_manager.go:597 +0xd5
github.com/golang/dep/internal/gps.maybeGitSource.try(0xc4201b6500, 0xb186c0, 0xc420012278, 0xc42015af00, 0x1b, 0xb1b400, 0xc420208040, 0xc420067080, 0x0, 0x0, ...)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/maybe_source.go:92 +0x1fe
github.com/golang/dep/internal/gps.maybeSources.try(0xc42018a580, 0x4, 0x4, 0xb186c0, 0xc420012278, 0xc42015af00, 0x1b, 0xb1b400, 0xc420208040, 0xc420067080, ...)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/maybe_source.go:46 +0x104
github.com/golang/dep/internal/gps.(*maybeSources).try(0xc4201b03a0, 0xb186c0, 0xc420012278, 0xc42015af00, 0x1b, 0xb1b400, 0xc420208040, 0xc420067080, 0x30, 0x30, ...)
        <autogenerated>:1 +0xbb
github.com/golang/dep/internal/gps.(*sourceGateway).require(0xc420218000, 0xb186c0, 0xc420012278, 0x1, 0x40, 0x0, 0x0)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/source.go:569 +0x210
github.com/golang/dep/internal/gps.(*sourceGateway).sourceURL(0xc420218000, 0xb186c0, 0xc420012278, 0x0, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/source.go:537 +0xb5
github.com/golang/dep/internal/gps.(*sourceCoordinator).getSourceGatewayFor(0xc42016fab0, 0xb186c0, 0xc420012278, 0xc4201a8180, 0x1c, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/source.go:205 +0x9f7
github.com/golang/dep/internal/gps.(*SourceMgr).SyncSourceFor(0xc4200670e0, 0xc4201a8180, 0x1c, 0x0, 0x0, 0x0, 0x0)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/source_manager.go:453 +0x82
github.com/golang/dep/internal/gps.(*bridge).SyncSourceFor(0xc42018ce10, 0xc4201a8180, 0x1c, 0x0, 0x0, 0xc4201dc000, 0x3)
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/bridge.go:216 +0x5b
created by github.com/golang/dep/internal/gps.(*solver).selectRoot
        /home/travis/gopath/src/github.com/golang/dep/internal/gps/solver.go:601 +0x614
```

### What did you expect to see?
Not that - I've never used dep on one of my own projects before, but roughly "dependencies resolved!" or something like.

### What did you see instead?
The above panic.

I was hit by this today as well, I added some debug output to the top of the listVersions loop to see the lines that it tried to parse, and the cause seems to be a warning from git that causes my panic:
Line: warning: invalid credential line: get
panic: runtime error: slice bounds out of range

goroutine 50 [running]:
github.com/golang/dep/internal/gps.(*gitSource).listVersions(0xc42048a3a0, 0xb65f40, 0xc420308320, 0xc4200bd3b8, 0x4123e8, 0x10, 0x8a4dc0, 0xc4200bd301)
	/go/src/github.com/golang/dep/internal/gps/vcs_source.go:225 +0x129d
github.com/golang/dep/internal/gps.maybeGitSource.try.func1(0xb65f40, 0xc420308320, 0xb65c40, 0xc42026e080)
	/go/src/github.com/golang/dep/internal/gps/maybe_source.go:97 +0x71
github.com/golang/dep/internal/gps.(*supervisor).do(0xc42027e1e0, 0xb65c40, 0xc42031a780, 0x9144f5, 0xc, 0x1, 0xc4200bd508, 0xc42028e0a0, 0x0)
	/go/src/github.com/golang/dep/internal/gps/source_manager.go:615 +0xd5
github.com/golang/dep/internal/gps.maybeGitSource.try(0xc420096c80, 0xb65c40, 0xc42031a780, 0xc420277880, 0xb, 0xb68960, 0xc420374240, 0xc42027e1e0, 0xc420306400, 0x0, ...)
	/go/src/github.com/golang/dep/internal/gps/maybe_source.go:95 +0x1fe
github.com/golang/dep/internal/gps.maybeSources.try(0xc420096d80, 0x7, 0x8, 0xb65c40, 0xc42031a780, 0xc420277880, 0xb, 0xb68960, 0xc420374240, 0xc42027e1e0, ...)
	/go/src/github.com/golang/dep/internal/gps/maybe_source.go:45 +0x2b9
github.com/golang/dep/internal/gps.(*maybeSources).try(0xc42024a140, 0xb65c40, 0xc42031a780, 0xc420277880, 0xb, 0xb68960, 0xc420374240, 0xc42027e1e0, 0x30, 0x30, ...)
	<autogenerated>:1 +0xbb
github.com/golang/dep/internal/gps.(*sourceGateway).require(0xc420061c20, 0xb65c40, 0xc42031a780, 0x1, 0x40, 0x0, 0x0)
	/go/src/github.com/golang/dep/internal/gps/source.go:569 +0x210
github.com/golang/dep/internal/gps.(*sourceGateway).sourceURL(0xc420061c20, 0xb65c40, 0xc42031a780, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/golang/dep/internal/gps/source.go:537 +0xb5
github.com/golang/dep/internal/gps.(*sourceCoordinator).getSourceGatewayFor(0xc4202a44d0, 0xb65c40, 0xc42031a780, 0xc420023800, 0x30, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/golang/dep/internal/gps/source.go:205 +0x9f7
github.com/golang/dep/internal/gps.(*SourceMgr).ExportProject(0xc42027e240, 0xb65c40, 0xc42031a780, 0xc420023800, 0x30, 0x0, 0x0, 0xb68660, 0xc42024af80, 0xc420302780, ...)
	/go/src/github.com/golang/dep/internal/gps/source_manager.go:474 +0x84
github.com/golang/dep/internal/gps.WriteDepTree.func1.1(0xc42035c060, 0xb65c40, 0xc42031a780, 0xc420023800, 0x30, 0x0, 0x0, 0xb689e0, 0xc4202a3040, 0xc420023980, ...)
	/go/src/github.com/golang/dep/internal/gps/result.go:95 +0x382
github.com/golang/dep/internal/gps.WriteDepTree.func1(0x8, 0x932330)
	/go/src/github.com/golang/dep/internal/gps/result.go:110 +0xe5
github.com/golang/dep/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1(0xc42031a7c0, 0xc4203404e0)
	/go/src/github.com/golang/dep/vendor/golang.org/x/sync/errgroup/errgroup.go:58 +0x57
created by github.com/golang/dep/vendor/golang.org/x/sync/errgroup.(*Group).Go
	/go/src/github.com/golang/dep/vendor/golang.org/x/sync/errgroup/errgroup.go:55 +0x66


it turned out I had a bad "credential.helper" config set in my git config which was the cause of it.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,7 @@ BUG FIXES:
 * Releases targeting Windows now have a `.exe` suffix (#1291).
 * Adaptively recover from dirty and corrupted git repositories in cache (#1279).
 * Suppress git password prompts in more places (#1357).
+* Validate `git ls-remote` output and ignore all malformed lines (#1379)
 
 IMPROVEMENTS:
 

--- a/gps/vcs_source.go
+++ b/gps/vcs_source.go
@@ -10,6 +10,7 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"regexp"
 	"strings"
 
 	"github.com/Masterminds/semver"
@@ -116,6 +117,10 @@ func (bs *baseVCSSource) exportRevisionTo(ctx context.Context, r Revision, to st
 
 	return fs.CopyDir(bs.repo.LocalPath(), to)
 }
+
+var (
+	gitHashRE = regexp.MustCompile(`^[a-f0-9]{40}$`)
+)
 
 // gitSource is a generic git repository implementation that should work with
 // all standard git remotes.
@@ -238,6 +243,10 @@ func (s *gitSource) exportRevisionTo(ctx context.Context, rev Revision, to strin
 	return nil
 }
 
+func (s *gitSource) isValidHash(hash []byte) bool {
+	return gitHashRE.Match(hash)
+}
+
 func (s *gitSource) listVersions(ctx context.Context) (vlist []PairedVersion, err error) {
 	r := s.repo
 
@@ -298,6 +307,13 @@ func (s *gitSource) listVersions(ctx context.Context) (vlist []PairedVersion, er
 	vlist = make([]PairedVersion, len(all))
 	for _, pair := range all {
 		var v PairedVersion
+		// Valid `git ls-remote` output should start with hash, be at least
+		// 45 chars long and 40th character should be '\t'
+		//
+		// See: https://github.com/golang/dep/pull/1160#issuecomment-328843519
+		if len(pair) < 45 || pair[40] != '\t' || !s.isValidHash(pair[:40]) {
+			continue
+		}
 		if string(pair[41:]) == "HEAD" {
 			// If HEAD is present, it's always first
 			headrev = Revision(pair[:40])

